dask/array/optimization.py

```diff
@@ -206,3 +211,60 @@ def fuse_slice(a, b):
             j += 1
         return tuple(result)
     raise NotImplementedError()
+
+
+def getitem_broadcast(arr, ind):
+    if not hasattr(arr, 'shape') or arr.shape == ():
+        return arr
+    ndim = len(arr.shape)
+    if ndim < len(ind):
+        ind = ind[-ndim:]
+    ind = tuple(i if s > 1 else 0 if isinstance(i, int) else slice(None)
+                for (i, s) in zip(ind, arr.shape))
+    return arr[tuple(ind)]
+
+
+# Create a dict of {func: nargs}
+binops = ('lt le eq ne ge gt add and_ floordiv lshift mod or_ pow rshift sub '
+          'truediv xor').split()
+PY2 and binops.append('div')
+func_lookup = dict.fromkeys((getattr(operator, o) for o in binops), 2)
+func_lookup.update((getattr(operator, o), 1) for o in ('invert', 'neg', 'pos'))
+
+
+def iselemwise(f):
+    return hasattr(f, 'nin') and hasattr(f, 'ntypes') or f in func_lookup
+
+
+def getargs(task):
+    """Get array arguments from an elementwise tasks"""
+    func = task[0]
+    if hasattr(func, 'nin'):
+        n = func.nin
+    elif func in func_lookup:
+        n = func_lookup[func]
+    else:
+        raise ValueError('Unknown function {0}'.format(func))
+    return task[1:n + 1]
+
+
+def swap_getitem_elemwise(dsk, keys):
+    dsk2 = dict()
+    for k, v in dsk.items():
+        try:
+            if (istask(v) and v[0] == getitem and v[1] in dsk and
+                    istask(dsk[v[1]]) and iselemwise(dsk[v[1]][0]) and
+                    isinstance(v[2], tuple)):
+                elemwise_task = dsk[v[1]]
```

Comment:
Thoughts on repeated application as in `((x + 1) * 2)[:5]`?  Perhaps this needs another loop?